File: Machines/Machine.py
from Machines.PinNode import PinNode
import os
import webbrowser

class Machine:
    def __init__(self):
        self.first = None
        self.last = None
        self.index = 0

    def insert(self,listElements):
        if self.first:
            self.last.next = PinNode(self.index,listElements)
            self.last.prev = self.last
            self.last = self.last.next
            self.index += 1
            return
        self.first = PinNode(self.index,listElements)
        self.last = self.first
        self.index += 1

    def iterated(self):
        current = self.first
        while current:
            print('Pin:',current.index)
            current.listElements.iterated()
            current = current.next

    def getPin(self,index):
        current = self.first
        while current:
            if current.index == index:
                return current
            current = current.next

    def size(self):
        return self.index
    
    def sizeElements(self):
        return self.first.index
    
    #Graphviz
    def getDot(self):
        dot = 'digraph maquina {\nnode1 [shape=none, margin=0, label=\n<<TABLE BORDER="0" CELLBORDER="0" CELLSPACING="0" CELLPADDING="5">\n'

        current = self.first
        while current:
            dot += '<tr>\n'
            dot += f'<td BGCOLOR="gray" width="60" height="30">Pin {current.index}</td>\n'

            currentE = current.listElements.first
            while currentE:
                dot += f'<td BGCOLOR="{currentE.element.color}" width="100" height="30">{currentE.element.atomcNum}<br align="left"/>{currentE.element.symbol}<br/>{currentE.element.name}</td>\n'
                currentE = currentE.next
            dot += '</tr>\n'
            current = current.next

        dot += '</TABLE>>\n'
        dot += '];\n'
        dot += '}'
        #print(dot)

        with open('Img/imgMachine.txt','w',encoding='utf-8') as report:
            report.write(dot)

        os.system('dot -Tjpg Img/imgMachine.txt -o Img/imgMachine.jpg')
        webbrowser.open('Img\imgMachine.jpg')

File: Frontend/js/machine.js
dot = d3.select('#machine').graphviz().scale(2.3).height(document.getElementById('machine').clientHeight).width(800*1.9).renderDot(`digraph maquina {
    node1 [shape=none, margin=0, label=
    <<TABLE BORDER="0" CELLBORDER="0" CELLSPACING="0" CELLPADDING="5">
    <tr>
    <td BGCOLOR="gray" width="60" height="30">Pin 1</td>
    <td BGCOLOR="#EAE37B" width="100" height="30">3<br align="left"/>Li<br/>Litio</td>
    <td BGCOLOR="#F675BA" width="100" height="30">2<br align="left"/>He<br/>Helio</td>
    <td BGCOLOR="#8F38CC" width="100" height="30">7<br align="left"/>N<br/>Nitrógeno</td>
    </tr>
    <tr>
    <td BGCOLOR="gray" width="60" height="30">Pin 2</td>
    <td BGCOLOR="#98ABA3" width="100" height="30">6<br align="left"/>C<br/>Carbono</td>
    <td BGCOLOR="#295F82" width="100" height="30">4<br align="left"/>Be<br/>Berilio</td>
    <td BGCOLOR="#40A90F" width="100" height="30">1<br align="left"/>H<br/>Hidrógeno</td>
    </tr>
    </TABLE>>
    ];
    }`)